pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/KafkaConnectSink.java

```diff
@@ -0,0 +1,277 @@
+        taskContext =
+                new PulsarKafkaSinkTaskContext(configs.get(0), task::open);
+        task.initialize(taskContext);
+        task.start(configs.get(0));
+
+        batchSize = kafkaSinkConfig.getBatchSize();
+        lingerMs = kafkaSinkConfig.getLingerTimeMs();
+        scheduledExecutor.scheduleAtFixedRate(() ->
+                this.flushIfNeeded(true), lingerMs, lingerMs, TimeUnit.MILLISECONDS);
+
+        isRunning = true;
+        log.info("Kafka sink started : {}.", props);
+    }
+
+    private void flushIfNeeded(boolean force) {
+        if (force || numPendingRecords.get() >= batchSize) {
+            scheduledExecutor.submit(this::flush);
+        }
+    }
+
+    public void flush() {
+        if (log.isDebugEnabled()) {
+            log.debug("flush requested, pending: {}, batchSize: {}",
+                    numPendingRecords.get(), batchSize);
+        }
+
+        if (numPendingRecords.getAndSet(0) == 0) {
+            return;
+        }
+
+        Map<TopicPartition, OffsetAndMetadata> currentOffsets = taskContext.currentOffsets();
+        CompletableFuture<Void> flushCf;
+        synchronized (this) {
+            flushCf = pendingFlush;
+            pendingFlush = new CompletableFuture<>();
+        }
+
+        try {
+            task.flush(currentOffsets);
+            taskContext.flushOffsets(currentOffsets);
+            flushCf.complete(null);
+        } catch (Throwable t) {
+            log.error("error flushing pending records", t);
+            flushCf.completeExceptionally(t);
+        }
+    }
+
+    /**
+     * org.apache.kafka.connect.data.Schema for the object
+     * @param obj
+     * @return org.apache.kafka.connect.data.Schema
+     */
+    public static Schema getKafkaConnectSchemaForObject(Object obj, Schema defaultSchema) {
+        if (obj == null) {
+            return defaultSchema;
+        }
+
+        if (primitiveTypeToSchema.containsKey(obj.getClass())) {
+            return primitiveTypeToSchema.get(obj.getClass());
+        }
+
+        // Other types are not supported yet.
+        // Will fallback to defaults provided.
+        return defaultSchema;
+    }
+
+    private SinkRecord toSinkRecord(Record<Object> sourceRecord) {
+        final int partition = 0;
+        final Object key;
+        final Object value;
+        if (unwrapKeyValueIfAvailable && sourceRecord.getValue() instanceof KeyValue) {
+            KeyValue<Object, Object> kv = (KeyValue<Object, Object>) sourceRecord.getValue();
+            key = kv.getKey();
+            value = kv.getValue();
+        } else {
+            key = sourceRecord.getKey().orElse(null);
+            value = sourceRecord.getValue();
+        }
+        final Schema keySchema = getKafkaConnectSchemaForObject(key, defaultKeySchema);
+        final Schema valueSchema = getKafkaConnectSchemaForObject(value, defaultValueSchema);
+
+        long offset = taskContext.currentOffset(topicName, partition).incrementAndGet();
```

Comment:
I am not sure if this is the correct logic. I believe this is the offset from the source partition. Hence, you should get the underlying Pulsar message, retrieve message-id and convert it to a Long number. You shouldn't assign an offset for it, which would result in an indeterministic behavior.

Reply:
@sijie The Record<> may or may not contain sequenceID (Optional) 
sequenceID, at the best, is `Optional.of(FunctionCommon.getSequenceId(message.getMessageId()))` - as i understand this applies to persistent topics only.

Record may or may not contain Message (Optional): sourceRecord.getMessage() which may return sequenceID or -1 (kafka has a thing against negative offsets, SinkRecord does not allow them).
MessageId can return byte[] (sourceRecord.getMessage().get().getMessageId().toByteArray()) of something - not guaranteed to be offset, rather a byte[] of anything, i.e. ledgerId, entryId, batchSize, topicName etc. Or it can throw UnsupportedOperationException

AFAICT usable sequenceID can be retrieved from data produced by kafka source (and there is absolutely no point in connecting kafka source to kafka sink in pulsar) or from the data read from bookkeeper. 

I can do something along the lines of
```
        long offset = sourceRecord.getRecordSequence()
                .orElse(taskContext.currentOffset(topicName, partition).incrementAndGet());
```

This assumes that topic either has sequenceID for all records, or does not have them for all records.

Next problem is that kafka's offsets increase monotonically (per partition).
In case of BK, offsets is `(ledgerId << 28) | entryId` and is not guaranteed to be sequential, as I understand. 
I don't know how kafka sinks react to that. Some sinks ignore offsets, others may track specific invariants.

Simply assigning the offsets the way I do now is the simplest solution. 
It won't allow Exactly-once processing if the kafka sink deduping/overwriting data by offset id (and not the key, and if it supports this) but it will work as AtLeast-once. 
I.e. if everything restarts it won't re-assign the same offsets (so deduping sink won't overwrite older data) but in case of replay/retry it will assign new offsets to the data potentially processed by the sink.

What do you hink?